knowledge_graph.py

```diff
@@ -0,0 +1,110 @@
+"""Knowledge Graph"""
+from typing import Dict, List, Any
+from transformers import (
+    Text2TextGenerationPipeline,
+    AutoModelForSeq2SeqLM,
+    AutoTokenizer)
+from transformers.pipelines.text2text_generation import ReturnType
+from pydantic import Field
+
+class KnowledgeGraph():
+  """Knowledge Graph class"""
+
+  max_new_tokens: int = Field(
+    default=200, description="Maximum number of new tokens to generate."
+  )
+  predictions : List[Dict[str, Any]] = Field(default=[])
+
+  def __init__(
+    self,
+    *args,
+    model_name:str='Babelscape/rebel-large',
+    **kwargs):
+
+    self.model = EntityRelationshipExtractionPipeline(model=model_name, *args, **kwargs)
+
+  def inference(self, sentence: str) -> Dict[str, Any]:
+    """inference"""
+    self.predictions = self.model(sentence)
+
+    return self.predictions
+
+class EntityRelationshipExtractionPipeline(Text2TextGenerationPipeline):
+  """Key phrase extractor pipeline class"""
+  def __init__(self, model, *args, **kwargs):
+    super().__init__(
+      model=AutoModelForSeq2SeqLM.from_pretrained(model),
+      tokenizer=AutoTokenizer.from_pretrained(model),
+      *args,
+      **kwargs)
+
+  def _extract_entity_relationships(self, text) -> List[Dict[str, Any]]:
+    triplets = []
+    relation = ""
+    subject = ""
+    relation = ""
+    object_ = ""
+    text = text.strip()
+    current = 'x'
+    for token in text.replace("<s>", "").replace("<pad>", "").replace("</s>", "").split():
+      if token == "<triplet>":
+        current = 't'
+        if relation != '':
+          triplets.append(
+            { "subject": subject.strip(), "relation": relation.strip(), "object": object_.strip() })
+          relation = ''
+        subject = ''
+      elif token == "<subj>":
+        current = 's'
+        if relation != '':
+          triplets.append(
+            { "subject": subject.strip(), "relation": relation.strip(), "object": object_.strip() })
+        object_ = ''
+      elif token == "<obj>":
+        current = 'o'
+        relation = ''
+      else:
+        if current == 't':
+          subject += ' ' + token
+        elif current == 's':
+          object_ += ' ' + token
+        elif current == 'o':
+          relation += ' ' + token
+    if subject != '' and relation != '' and object_ != '':
+      triplets.append(
+        { "subject": subject.strip(), 'relation': relation.strip(), 'object': object_.strip() })
+    return triplets
```

Comment:
<img src="https://avatars.githubusercontent.com/in/347564?s=41" alt="Image description" width="20" height="20">   CodeRabbit

The `_extract_entity_relationships` method lacks error handling. If the input text doesn't follow the expected format, this could lead to unexpected behavior or errors. Consider adding error handling or checks to ensure the input text is in the correct format.

Also, the method is quite complex and could benefit from refactoring for better readability and maintainability. For example, you could create helper methods for each 'current' state ('t', 's', 'o') to handle the logic for that state.

Lastly, there's a redundancy in the code where the same triplet appending operation is performed multiple times. This can be simplified by moving it into a separate function.

```diff
  def _extract_entity_relationships(self, text) -> List[Dict[str, Any]]:
    triplets = []
    relation = ""
    subject = ""
    relation = ""
    object_ = ""
    text = text.strip()
    current = 'x'
    for token in text.replace("<s>", "").replace("<pad>", "").replace("</s>", "").split():
      if token == "<triplet>":
        current = 't'
        if relation != '':
-            triplets.append(
-              { "subject": subject.strip(), "relation": relation.strip(), "object": object_.strip() })
+            self._append_triplet(triplets, subject, relation, object_)
          relation = ''
        subject = ''
      elif token == "<subj>":
        current = 's'
        if relation != '':
-            triplets.append(
-              { "subject": subject.strip(), "relation": relation.strip(), "object": object_.strip() })
+            self._append_triplet(triplets, subject, relation, object_)
        object_ = ''
      elif token == "<obj>":
        current = 'o'
        relation = ''
      else:
        if current == 't':
          subject += ' ' + token
        elif current == 's':
          object_ += ' ' + token
        elif current == 'o':
          relation += ' ' + token
    if subject != '' and relation != '' and object_ != '':
-      triplets.append(
-        { "subject": subject.strip(), 'relation': relation.strip(), 'object': object_.strip() })
+      self._append_triplet(triplets, subject, relation, object_)
    return triplets
+
+  def _append_triplet(self, triplets, subject, relation, object_):
+    """Appends a new triplet to the list if all elements are not empty."""
+    if subject and relation and object_:
+      triplets.append(
+        { "subject": subject.strip(), "relation": relation.strip(), "object": object_.strip() })
```


<!-- This is an auto-generated comment by OSS CodeRabbit -->